src/cachemachine/automatedteller.py

```diff
@@ -0,0 +1,75 @@
+import asyncio
+from typing import Any, Dict, List, Sequence, Union
+
+import structlog
+
+from cachemachine.cachechecker import CacheChecker
+from cachemachine.cachedepositer import CacheDepositer
+from cachemachine.rubinrepoman import RubinRepoMan
+from cachemachine.simplerepoman import SimpleRepoMan
+
+logger = structlog.get_logger(__name__)
+
+
+class AutomatedTeller:
+    available_images: List[Dict[str, str]] = []
+    desired_images: List[Dict[str, str]] = []
+    images_to_cache: List[Dict[str, str]] = []
```

Comment:
What might be helpful is to think of regular Python classes and dataclasses as two completely different things. They aren't, quite, but it's best to treat them that way.

In a regular Python class (and all of this is unrelated to typing; this is always the case for any Python class), class variables are defined in the class (outside of any method), and instance variables are defined as attributes of `self` inside a method. The bit that typing adds is that if you want to use instance variables, all of your instance variables must be defined inside `__init__` so that typing knows about them. But other than that, typing doesn't change anything.

You can just move those variables into `__init__` and everything should work the same (except more robustly).

Dataclasses are very, very special in part because they have to wedge themselves into Python standard syntax. In a dataclass, the dataclass library introspects the class and uses all the class variables as a *template* to define each instance in the auto-generated constructor. So every class variable will get a corresponding instance variable in each instance that shadows the class variable. You never want to use the class variables in dataclasses; they're just an implementation and syntax artifact.

Therefore, in dataclasses *only* you define all the instance variables at the top level as if they're class variables and then the library does a bunch of magic. But for any normal class, instance variables go in `__init__`.

Reply:
Okay will move the variables into init, but I think I got confused by this:

https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html

Looking at the Classes section, it basically shows that the way that I'm doing it isn't a classvariable, because I'm not using ClassVar.  Even though it's in the class and looks like a class variable. Although maybe this is outdated or something, I honestly can't tell what the right PEP is anymore.

Reply:
Ah, that summary gives the wrong impression. The full documentation is clearer, I think:

https://mypy.readthedocs.io/en/stable/class_basics.html

The key point, I think, is that typing doesn't change anything about Python itself works. It just gives mypy enough information to issue diagnostics. So if something is a class variable normally in Python, it still will be when using annotations (and the other way around).

`ClassVar` is a type hint for mypy intended to solve a different problem, which is that it's easy to override class variables by mistake. Consider the following valid Python code:

```
class SomeClass:
    variable = True

    def method(self):
        self.variable = False
```
This code doesn't change the class variable to `False` as one might expect. Instead, it creates a new instance variable that shadows the class variable and sets it to `True`. This can be very confusing, but it's also used intentionally in Python so it can't just be rejected with a diagnostic. The purpose of the explicit `ClassVar` type hint is to work around this and tell mypy "this is just a class variable and there will be no shenanigans, so if I accidentally create an instance variable of the same name, throw an error." If one adds a `ClassVar` annotation to the first setting of `variable` above, mypy will throw an error inside `method` and tell you to use `SomeClass.variable` instead.

But `ClassVar` only affects mypy, not Python, so it doesn't change whether or not something is a class variable. If it's declared outside of a method body, it's a class variable, with or without `ClassVar`, and thus is effectively a global.

Reply:
I see, my cheat sheet is wrong apparently, but god is that confusing.  It says instance variable right there!  GAAAAAAAH.  I will vent my frustration at python's typing system at 1:30!